### DISCUSSION TITLE
How to upgrade miniforge for virtual environment?

### DISCUSSION THREAD

**Discussion by lamrongol** [conda-forge/miniforge#650] [Q&A] [answered]
When installing new version of miniforge, I want to upgrade them in virtual environment. How can I upgrade?
A way I thought is `conda list --export > requirements.txt` and upgrade miniforge and re-create virtual environment `conda create --name <env> --file requirements.txt`.

**Comment by hmaarrfk** [comment]
just type
```
mamba install "package_name>=x.y.z"
```
where x.y.z is the version of the package you want.

```
mamba update --all
```
can work too but it is ambiguous as to which package should take priority.

**Comment by lamrongol** [comment]
I want to upgrade `python.exe` itself. I found followings are similar question:
- [pip - Upgrade Python in a virtual environment - Stack Overflow](https://stackoverflow.com/questions/10218946/upgrade-python-in-a-virtual-environment)
- [How to change Python version of existing conda virtual environment? - Stack Overflow](https://stackoverflow.com/questions/59163078/how-to-change-python-version-of-existing-conda-virtual-environment)

But I don't know they are applicable to miniforge.

**Comment by hmaarrfk** [comment]
so in this case your package is "python"
```
mamba install "package_name>=x.y.z"
```
becomes

```
mamba install "package_name=3.12"
```

(its better to just pin to 3.12, 3.13 is still very very new).

**Comment by lamrongol** [comment]
This may be my misremembering but miniforge(pypy3) compiles `python.exe` in their own way, this is just my case, I use PyCharm and create virtual environment through PyCharm, environment folder is created in miniforge folder and when installing new version of miniforge, deleting existing environments are needed.
For miniforge, does `conda install "python=3.9"` work and shouldn't I use PyCharm to create virtual environments?

**Reply by hmaarrfk** [reply]
so virtual environments and conda environments are related but different technologies.

When you use "conda environments", a new python installation is made within the environment folder. Everything refers to the inside of the newly created environment as much as possible.

When you use "python's virtual environment", you continue to refer to the "original" python that created it.

so when you update the original python, the packages that you instealled in your virtual environment get "lost". They thought they were supposed to target python 3.9, but all of a sudden, you are using 3.12.

conda and pip/venv don't talk to each other as much as you think they do.

See https://www.anaconda.com/blog/using-pip-in-a-conda-environment for details and recommendations

**Reply by lamrongol** [reply]
To be honest, I'm not so familiar with Python and conda, I want to use machine learning packages such as `pandas` on PyPy and try to use miniforge-pypy3.
The immediate remedy is to re-create "conda environment" by using `conda list --export > requirements.txt`, is it right?

**Reply by hmaarrfk** [reply]
Unfortunately "machine learning packages" + pypy is very difficult and mostly abandoned
Please see: https://conda-forge.org/news/2024/08/14/sunsetting-pypy/

I don't think it exists on pypi, nor does it seem to exist on conda-forge
https://pypi.org/project/pandas/#files   (if you cp = CPython, and there is no "pypy" in the filenames).
https://anaconda.org/conda-forge/pandas/files

Good luck with finding the right mix of dependencies for PyPy.

I would recommend if you are getting started to use CPython. Later on, you can experiment with PyPy.

**Reply by lamrongol** [reply]
I've heard that not pypy3 but "miniforge-pypy3" can use `pandas`and `sckit-learn` and so on. However, now I searched and `numpy` and `scipy` are fairly good enough for machine learning.
I'm going to start trying those.

**Comment by hmaarrfk** [comment, marked as answer]
Generally speaking, i don't "upgrade" my environments. 

I just "delete them" and recreate them.

Goodluck.

**Comment by lamrongol** [comment]
I had been mistaken for a long time but machine learning packages such as `sckit-learn` are originally accelerated and additional "acceleration" is not needed.
Maybe there are other people who have same mistaken, so I wrote.

**Reply by hmaarrfk** [reply]
Oh. You aren’t alone. And while it is true that they are “accelerated” the algorithms can always go faster, but it depends how long you want to spend making that happen. 

Happy coding!
